[RNMobile] File Block II

etoledom opened the pull request:
<!-- Learn the overall process and best practices for pull requests at https://github.com/WordPress/gutenberg/blob/master/docs/contributors/repository-management.md#pull-requests. -->

## Description
<!-- Please describe what you have changed or added -->

This PR implements most of the options for File Block:
- Download button dimming while upload is in progress.
- "Open in new tab" option.
- "Show download button" option.
- "Copy file URL" option.
- Alignment rendering.
- "Edit file" button (choose a different file).
- Improved design (Font size/color)

For the copy file URL feature, I decided to close the bottom-sheet and show a notice, instead of keeping the bottom-sheet and changing the button text. @iamthomasbishop - let me know if you still prefer the later, it would be a simple change. 

<img src="https://user-images.githubusercontent.com/9772967/96729455-75206e00-13b5-11eb-8de5-5435b167950e.gif"  width="200">

## How has this been tested?

- Download button dimming while upload is in progress.
  - Upload a file
  - Check that the button is dimmed out while the file is uploading.
  - Keep in mind that on the Demo app, the upload is fake (iOS).
- "Open in new tab" option.
  - Open the block settings (`cog` button)
  - Switch `on` the "Open in new tab" option.
  - Switch to the HTML mode
  - Search for the `target="_blank" rel="noreferrer noopener"` attributes in the `<a>` tag.
- "Show download button" option.
  - Open the block settings (`cog` button).
  - Switch `off` the "Show download button` option.
  - Check that the button has disappear.
  - Switch to the HTML mode.
  - Check there's only one `<a>` tag.
- "Copy file URL" option.
  - Open the block settings (`cog` button).
  - Press on the `Copy file URL` option.
  - Check that the bottom-sheet hides out
  - Check that the notice `Copied` is displayed.
  - Paste the clipboard content somewhere else (another block or app).
  - Check that the full file URL is pasted (will be a local file URL for the Demo app).
- Alignment rendering.
  - Press on the Alignment toolbar button.
  - Choose the different options.
  - Check that the UI represents the alignment accordingly.
  - Note: `Alight left`, `Wide width` and no option selected won't have visual differences.
- "Edit file" button (choose a different file).
  - Press the Edit file button (the one with the arrows).
  - Check that you are presented with the Media Sources option.
  - Check that you are able to choose and update the block with a new file.


## Screenshots <!-- if applicable -->

## Types of changes
<!-- What types of changes does your code introduce?  -->
<!-- Bug fix (non-breaking change which fixes an issue) -->
<!-- New feature (non-breaking change which adds functionality) -->
<!-- Breaking change (fix or feature that would cause existing functionality to not work as expected) -->

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/javascript/ -->
- [x] My code follows the accessibility standards. <!-- Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/accessibility/ -->
- [x] My code has proper inline documentation. <!-- Guidelines: https://developer.wordpress.org/coding-standards/inline-documentation-standards/javascript/ -->
- [x] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [x] I've updated all React Native files affected by any refactorings/renamings in this PR. <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/master/docs/contributors/native-mobile.md -->

etoledom committed:
Dim Download button when upload is in progress

etoledom committed:
Implementing replace file toolbar button

etoledom committed:
Implement Hide Download Button switch

etoledom committed:
Implementing Copy Link option

etoledom committed:
Implementing "Open in new tab" setting

etoledom committed:
Implement Alignment render of block elements

etoledom committed:
Small design improvements

etoledom committed:
Showing block toolbar buttons only when the update finishes

iamthomasbishop commented:
> For the copy file URL feature, I decided to close the bottom-sheet and show a notice, instead of keeping the bottom-sheet and changing the button text

@etoledom I think I'd prefer the inline feedback messaging, mainly because it's direct and much more obvious (the compact notice is way on the top of the screen and pretty subtle). I figured this could be a tricky thing when I was designing it, but it would be a nice subtle touch, and hopefully one that we can re-use elsewhere 😄

jd-alexander reviewed (commented):
Thanks for these changes @etoledom I reviewed the code and all the test scenarios described and they all worked as expected. I will do another pass once you have addressed the design feedback. 👍

etoledom committed:
Dont hide bottom sheet after copying url

etoledom committed:
Styles refinements:
- Copy URL action button color
- Media Placeholder icon color (BlockIcon mod)

etoledom commented:
I have rebased and updated this branch after the merge of the [first iteration PRs ](https://github.com/WordPress/gutenberg/pull/26278)

Copy URL behaviour has been changed as @iamthomasbishop requested:

![Copy url](https://user-images.githubusercontent.com/9772967/99073773-cc7dbc80-25b6-11eb-9245-68b4eecfef99.gif)

Also, some minor style improvements:
- Media Placeholder icon color.
- Copy url action button color.

iamthomasbishop commented:
Looks awesome, @etoledom. Thanks for taking care of that!

jd-alexander commented on packages/block-library/src/file/edit.native.js:
```diff
@@ -70,6 +95,29 @@ export default class FileEdit extends Component {
+		this.setState( { isUrlCopied: true } );
+		this.timerRef = setTimeout( () => {
+			this.setState( { isUrlCopied: false } );
+		}, 1500 );
```
@etoledom is it okay to have this value hardcoded here or do we want to put it in a constant? Whichever works for me. WDYT?

etoledom replied:
Done in https://github.com/WordPress/gutenberg/pull/26360/commits/53a20651d130de3195faadbdddfb796782c4c091
Nice catch!

jd-alexander reviewed (approved):
These changes are looking good @etoledom I reviewed the code and tested all the flows and they worked as expected. I left a few minor comments which you can address as you see fit. LGTM 🚢

jd-alexander commented on packages/block-library/src/file/edit.native.js:
```diff
@@ -138,34 +262,87 @@ export default class FileEdit extends Component {
+							{ getMediaOptions() }
+							{ this.getInspectorControls( attributes ) }
 							<RichText
 								__unstableMobileNoFocusOnMount
```
@etoledom just a thought here. Do you think we should disable the user's ability to edit the filename while the upload is taking place? I was just doing an upload and when I realized I could still interact with it, it felt like something that should be done when we are sure the file upload is successful. Looking forward to hearing your thoughts on this.

etoledom replied:
Good catch!
I just tested on the Web version and we are able to edit the file name also there. Since it's no more than an attribute change I don't think it's a big deal (unless there's any related bug/UX issue with this).

I personally tend to prefer copying what's in Web, but not hard feelings.
@iamthomasbishop - any thought about this?

etoledom replied:
What we should definitely block is the Copy URL button.
Done in https://github.com/WordPress/gutenberg/pull/26360/commits/e777c4b8ed77cc04a647d92effa5c5f1d114720c

etoledom replied:
![disable_button](https://user-images.githubusercontent.com/9772967/99256210-59be4c80-2815-11eb-9ac6-15ec9208ab6b.gif)

cc @iamthomasbishop

jd-alexander replied:
> What we should definitely block is the Copy URL button.

This sounds like a good idea 😄

iamthomasbishop replied:
> we should definitely block is the Copy URL button.

Can we set the button to a disabled state? Not sure if we already have a style for that. If not, we can probably set the opacity to 25%, which will end up being our standard `quaternary`/`disabled` style when our new color system is fully-implemented.

etoledom replied:
Thank you @iamthomasbishop ! Done in https://github.com/WordPress/gutenberg/pull/26360/commits/262475dec54632d810696e8ba7937047039e52b7

I also changed the Download button to 25% opacity when disabled.

![link_button](https://user-images.githubusercontent.com/9772967/99361927-1c100100-28b3-11eb-9e1b-e2505763cd85.gif)

iamthomasbishop replied:
@etoledom that's definitely better. One more tiny thing: can we also change the text-color (in the `disabled` state) to white?

etoledom committed:
Using constant for timer timeout

etoledom committed:
Disable Copy Link button while upload is in progress

jd-alexander reviewed (approved):
Thanks for these changes @etoledom The code looks good.

etoledom committed:
Dim buttons when they are dissabled

jd-alexander commented on packages/block-library/src/file/edit.native.js:
```diff
@@ -110,24 +160,113 @@ export default class FileEdit extends Component {
+								? __( 'Copied!' )
+								: __( 'Copy file URL' )
+						}
+						labelStyle={
+							this.state.isUrlCopied || finalButtonStyle
+						}
```
Hi @etoledom can you explain how this ||-operator evaluates here. What happens if this.state.isUrlCopied is true? I am going to test this shortly but I am just asking here since you might be able to expound clearly. I am just verifying that I understand how these operators work in Javascript :)

etoledom replied:
Sure, so basically the right hand side will be returned only if the left hand side is `false`. If the LHS is `true`, then `undefined` is returned.

This is a short for:
```js
return this.state.isUrlCopied ? undefined : finalButtonStyle;
```

The oposite is true using `&&` operator instead.

mchowning replied:
> If the LHS is true, then undefined is returned.

I thought if the LHS of `||` was truthy, that value would be returned, so if the LHS was `true` it would return `true`. Am I misunderstanding javascript (again 😄 )?

etoledom replied:
Wait a sec, you are right! It's not returning `undefined` but the LHS itself. This is new to me 🤔 
Thanks for clarifying @mchowning !

jd-alexander replied:
Just to clarify. So if it's returning the LHS then does that mean if url is copied then the style would be set to true instead of an actual style ? That's what I'm trying to clear up.

etoledom replied:
You are right, the style would be set to true.
I'm doubting this now myself, you know 🤣
But there are a few instances of this same technique used in styles in other parts of the project, so I'm inclined to say it's safe.

jd-alexander replied:
Okay understood! LGTM 🚢
